### ISSUE TITLE
Further camera improvements

### ISSUE THREAD

**Issue by aeickhoff** [citybound/citybound#85]
- [ ] Implement panning in a way that tries to preserve the 3D position of the cursor
(So when I start to pan and my cursor is over a specific point on the plane, after the pan the "world should be shifted" in a way to exactly align the new cursor position with the same 3D position - this should make the whole interaction feel much more physical and direct, like you're actually "grabbing and dragging" the world. Probably some 2D->3D and 3D->2D projections are needed for that. At least the first already exists in `monet`
- [ ] Smooth out camera movements
(Make camera movements appear smoother by controls not affecting the camera directly, but a camera target point instead, that the camera constantly interpolates towards, with a given interpolation "speed", which balances of directness vs. smoothness)

**Comment by aeickhoff** [comment]
Summoning @Herbstein ...

**Comment by Herbstein** [comment]
Should the smooth camera be applied to both the target and the position, or just the position?

**Comment by aeickhoff** [comment]
I meant: all controls directly influence the target position, rotation, etc. and all the actual camera ever does is approach its target (position, rotation)

**Comment by kingoflolz** [comment]
I think it might be difficult to apply interpolation, as the final and initial positions are not known ahead of time. Maybe it will be better to have the movement functions apply a "force" instead (capped at a max speed of course)?

**Comment by aeickhoff** [comment]
Yeah that actually captures better in words what I mean. Just something simple like `eye.position += (target.position - eye.position) * 0.5 * dt`

**Comment by kingoflolz** [comment]
I was thinking more of
```
if W{
    eye.velocity += dt * k;
    cap(eye.velocity, -1, 1);
}else{
    eye.velocity *= 0.99f; //not sure how to scale this one with dt
}

eye.position += eye.velocity;
```

**Comment by aeickhoff** [comment]
so the camera would accelerate if you hold a direction key? why?

**Comment by kingoflolz** [comment]
It would accelerate until it hits the max speed (maybe the speed we have right now), but I think it would feel smoother to the user.

**Comment by aeickhoff** [comment]
Ok, my suggestion of directly moving the target and interpolating the camera implicitly achieves the same, but you don't to maintain velocities and so on and also works for smoothing out position/rotation changes due to mouse movement - that's why I'm suggesting that. I implemented it like that in the old (JS) version of Citybound and really liked the result.

In the end, I only care about it feeling "right", I'll leave it up to you to experiment and implement it like you want.

**Comment by kingoflolz** [comment]
Sure, if it worked before, it will probably work again!

**Comment by Herbstein** [comment]
I've been looking for a place that either stores the the current mouse position or the mouse world position, but haven't found it anywhere.

Also, is there a reason for stuff like input not being globally accessible? It seems that it would alleviate some annoyances.

**Comment by aeickhoff** [comment]
That should actually always be updated, also when a modifier is down.

Once you fixed that, you should be able to use `UserInterface.cursor_2d` and `UserInterface.cursor_3d`.

**Comment by hubble618** [comment]
Zooming happens at the center of the screen, not the mouse. An option is needed to be able zoom where the mouse is pointed at.
